handlers/stream_cache.go

```diff
@@ -0,0 +1,123 @@
+package handlers
+
+import (
+	"fmt"
+	"sync"
+)
+
+// Shared server state. Each pipeline (usually endpoint) having separate books.
+// State is protected by mutex allowing concurent usage.
+// All state manipulation is contained in this file with goal to be brief and release mutex asap.
+type StreamCache struct {
+	Segmenting  SegmentingCache
+	Transcoding TranscodingCache
+}
+
+// Returns pointer so each handler would refer to same object (kind of singleton)
+func NewStreamCache() *StreamCache {
+	c := &StreamCache{}
+	c.Init()
+	return c
+}
+
+type TranscodingCache struct {
+	pushes map[string]*SegmentInfo
+	mutex  sync.Mutex
+}
+
+type SegmentInfo struct {
+	CallbackUrl   string
+	Source        string   // S3 input we are transcoding
+	UploadDir     string   // S3 destination url for multiple renditions
+	Destionations []string // Rendition URLS go here on push start and removed on push end
```

Comment:
Typo

Reply:
1. typo to fix
2. Kind of no, let me explain

Note that one input segment is transcoded into multiple rendition segments. We use single `UploadDir` as base for rendition URLs. Each rendition gets unique name and final URL is stored as record in `Destinations` when upload starts. On upload end we remove record from `Destinations` and finally on `Destinations` empty we remove `SegmentInfo` struct.

Reply:
Ok, then it makes sense!